### ISSUE TITLE
model.test_on_batch metric looks wrong after calling model.evaluate 

### ISSUE THREAD

**Issue by ForeverZyh** [keras-team/keras#14086]
<em>Please make sure that this is a Bug or a Feature Request and provide all applicable information asked by the template.
If your issue is an **implementation question**, please ask your question on [StackOverflow](http://stackoverflow.com/questions/tagged/keras) or [on the Keras Slack channel](https://keras-slack-autojoin.herokuapp.com/) instead of opening a GitHub issue.</em>  

**System information**  
- Have I written custom code (as opposed to using example directory):   Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):   Linux Ubuntu 18.04
- TensorFlow backend (yes / no):  yes
- TensorFlow version:  1.15.0
- Keras version:  2.3.1
- Python version:  3.6.8
- CUDA/cuDNN version: 10.2 
- GPU model and memory:  v100 32510MiB

You can obtain the TensorFlow version with:  
python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"  
You can obtain the Keras version with:  
python -c 'import keras as k; print(k.__version__)'  

**Describe the current behavior**  
Ex1:
```
# I have added a metric of accuracy
model.compile(optimizer=Adam(learning_rate=self.lr), loss='categorical_crossentropy', metrics=['accuracy'])
# ...
loss, acc = model.evaluate(test_X, test_Y) # call evaluate before test_on_batch
loss_b1, acc_b1 = model.test_on_batch(batch_X, batch_Y)
```
Ex2:
```
# I have added a metric of accuracy
model.compile(optimizer=Adam(learning_rate=self.lr), loss='categorical_crossentropy', metrics=['accuracy'])
# ...
# loss, acc = model.evaluate(test_X, test_Y) do not call evaluate before test_on_batch
loss_b2, acc_b2 = model.test_on_batch(batch_X, batch_Y)
```
Ex3:
```
# I have added a metric of accuracy
model.compile(optimizer=Adam(learning_rate=self.lr), loss='categorical_crossentropy', metrics=['accuracy'])
# ...
loss, acc = model.evaluate(test_X, test_Y) # call evaluate before evaluate
loss_b3, acc_b3 = model.evaluate(batch_X, batch_Y)
```
The `loss_b1`, `loss_b2`, `loss_b3` are the same. However, `acc_b2=acc_b3` but `acc_b1` is different from `acc_b2`,`acc_b3`.
**Describe the expected behavior**  
I would expect the `acc_b1=acc_b2=acc_b3`
**Code to reproduce the issue**  
Provide a reproducible test case that is the bare minimum necessary to generate the problem.  

**Other info / logs**  
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.

**Comment by ForeverZyh** [comment]
update Ex4:
```
# I have added a metric of accuracy
model.compile(optimizer=Adam(learning_rate=self.lr), loss='categorical_crossentropy', metrics=['accuracy'])
# ...
loss, acc = model.evaluate(test_X, test_Y) # call evaluate before test_on_batch
loss_b4, acc_b4 = model.test_on_batch(batch_X, batch_Y)
loss_b5, acc_b5 = model.test_on_batch(batch_X, batch_Y)
```

`acc_b5  = acc_b2 = acc_b3`, `acc_b4 = acc_b1`. I would imagine the `evaluate` method does not clear the `accuracy` metric.

**Comment by goelharshit562** [comment]
I have run all the experimets but all of my accuraccy  and loss is same.

**Comment by ForeverZyh** [comment]
@goelharshit562 Thanks. I try to minimize the reproducible code.

**Comment by ForeverZyh** [comment]
```
import keras
from keras.layers import Embedding, Input, Dense, Conv1D, Flatten
from keras.models import Model
from keras.utils import to_categorical
from keras.optimizers import Adam
import numpy as np


class char_AG:
    def __init__(self, lr=0.0009, all_voc_size=56, D=64):
        self.all_voc_size = all_voc_size
        self.D = D
        self.c = Input(shape=(300,), dtype='int32')
        self.y = Input(shape=(4,), dtype='int32')
        self.embed = Embedding(self.all_voc_size, self.D, name="embedding")
        look_up_c_d3 = self.embed(self.c)
        self.conv1d = Conv1D(64, 10, activation="relu")
        self.lr = lr
        x = self.conv1d(look_up_c_d3)
        x = Flatten()(x)
        self.fc1 = Dense(64, activation='relu')
        x = self.fc1(x)
        self.fc3 = Dense(4, activation='softmax')
        self.logits = self.fc3(x)
        self.model = Model(inputs=self.c, outputs=self.logits)
        self.model.compile(optimizer=Adam(learning_rate=self.lr), loss='categorical_crossentropy', metrics=['accuracy'])
        self.early_stopping = keras.callbacks.callbacks.EarlyStopping(monitor='val_loss', min_delta=0, patience=5,
                                                                      verbose=0, mode='auto',
                                                                      baseline=None, restore_best_weights=True)


def test_model():
    training_X = np.random.randint(0, 40, [100, 300])
    training_y = np.random.randint(0, 4, [100])
    test_X = np.random.randint(0, 40, [10, 300])
    test_y = np.random.randint(0, 4, [10])
    nb_classes = 4
    training_Y = to_categorical(training_y, nb_classes)
    test_Y = to_categorical(test_y, nb_classes)

    model = char_AG()
    normal_loss, normal_acc = model.model.evaluate(test_X, test_Y, batch_size=64, verbose=0)
    print("normal loss: %.4f\t normal acc: %.4f" % (normal_loss, normal_acc))

    loss, acc = model.model.test_on_batch(training_X[:64], training_Y[:64])
    print(loss, acc)
    loss, acc = model.model.test_on_batch(training_X[:64], training_Y[:64])
    print(loss, acc)
    loss, acc = model.model.test_on_batch(training_X[:64], training_Y[:64])
    print(loss, acc)


test_model()
```


Please try this code.

**Comment by ForeverZyh** [comment]
It does not work on CPU neither.

**Comment by ogrisel** [comment]
I have open a similar issue with tensorflow 2.4.1 on the tensorflow github tracker with a minimal reproducer:

https://github.com/tensorflow/tensorflow/issues/47340

**Comment by bhack** [comment]
It is fixed at https://github.com/keras-team/keras/pull/15342

**Comment by jvishnuvardhan** [comment]
Thanks @bhack. Still reproducing as the PR was not merged yet. I will check and close the issue once the PR gets merged. Thanks!

**Comment by ForeverZyh** [comment]
Thanks @bhack !

**Comment by jvishnuvardhan** [comment]
@ForeverZyh This is resolved by @bhack's PR.

I am closing this issue as this was resolved. Thanks @bhack

**Comment by bhack** [comment]
@qlzh727 Do you think we have another edge case not covered by the new test at: 

https://discuss.tensorflow.org/t/meanmetricwrapper-produces-inconsistent-results-on-multiple-runs/4714
